src/DIRAC/Resources/Computing/cloudinit.template

```diff
@@ -0,0 +1,125 @@
+      vLIlkkVzgHl54yc8tIkByWx64jNeDH7pkrRwfQMZ2RaPBoMiysdcxTR5gpmGDKewWpKm4Hpy5Df+
+      zEjEAANmhNHuEmhVMkU4yguukUUftuaiW0HgNBtFdcEDTI7eQ00UfHq7U22hgw2DrKSjIUkil8I9
+      03S+ux2zgWEdOcHUaGnTDT4yIEkKOLyaCH+cTGjnQ/8qZz95bkn3rVi/L4kPsmJ2Ptbl9J5ccRF5
+      Dz1H5n64xDMXAkWwMCS2SjdRB1aDGvcLlj6CwVXsrZSGRB376y3syRCNZhYEWAjbrJmNU8nrcUtg
+      6Clh3RI/IoSbqWjDIef5PuHueqx9L33bhIOPI829jcTgyI0bpljmbxTUdoXfm4WcpXLKXx/nHtzn
+      sUaHwMrQQIhl8LGrYctt+9g3GcPvNfJXZeuU421ZY444bjBnlsXKiR30g/b9hk8TuG7D0NxJQ1Te
+      3uibnbe9JaVW1NZFmroWGJaWL+ErR93vTtAcErKNJMA3f6r5Jgupf2RPYpYyCodcGMcblOBuWJyb
+      Mx9dwTDE0540Be+psuPT5DxXdnDIgySD6Q5IseFnFvYL90T2oWkgYco+ImeNJTvyxp2+VZf/VFGd
+      X7bulZ0VoXigGreq+4/RRwsLlavXy+OPonN0dSqtEtp4W8ged3a0eiN+8Mvce+qb2y9aKieZWBTB
+      FNLV1MMvydmafY3BZd/wXcPrAWxgARpamY8QB/ae9Vxl+zTvMuCrX21lCv4RByeIe79NUfmTifKe
+      9H6TM/ivN3vUloTRtEZDoAI1bcnO3Ncc+CwmXS0yAzIWVJHPAElmqoXjK4kFllT8MqwAUBibnlSC
+      qHXnbCE8KwX2iS3jNC5us7IaOY3XqqiiMm/xum9BPa24Xe4udRyU4Hm8FmNxLITY4IUVRk1QfbJq
+      cgX2O8+F54K3l1GjDLyFT6Ie7iS67OpBf0g4nSfFU14Y/6Ln2awRqkiUUngIG6Txa5jTApKQxmr2
+      B3ptSj9R3PjbPtybBIcrTr47tjPJwSpkgDhs53mwdyhzujjicwFR0tKjTHTl4wULqg3w44uVJo1v
+      au59VM5j9w+iKM9nqTVNgn5h9eHo91uHEza+BhycEONRNo1yYuYWGV99L3zr7Yai6HguUvulGw+s
+      oud6l8p8Z7FS1kudFsPj44gUYARG3yFxCP1GtHiqFgMwKq0tFZ83PyMDnq+lzPGkw4VvIQ1JPl2C
+      L0K1oTb7lT6J+sUO5e8O/K+5pZWVr3Ist34M2kScte0NrvVwCcTCCn3qZemVBGls6RLeBNGUEbay
+      MQrx5KnfFtYP71KVbLzIx2VvKuHtyHOvbt9+2UJ++KpydF4jmKCJym5krCPQAHDOeqc284pMMYO4
+      2ZlTW4CPRiqBje7b8jIc40TPWIzHkC97y11Yk/xNNNTbp942Kh88OGXseU6FoWPQzpi8RxiLi3aB
+      9T7IusO/n4jCj35rypno2jfVcrecG+6dDSRLp968oShjaYISo9y+QnazMCTPpir8IqfTecQ+eIpC
+      e5X1hmVj5b99b0inDi9A/HipHcEef/5Afz65mUD/+w5OZ//PsfwDIDTlCX4GAAA=
+ - path: /etc/cvmfs/default.local
+   content: |
+     CVMFS_CACHE_BASE=/mnt/cvmfs
+     CVMFS_HTTP_PROXY="DIRECT"
+ - path: /root/proxy.pem
+   permissions: '0600'
+   content: |
+     %(proxy)s
+ - path: /root/run_pilot.sh
+   permissions: '0755'
+   content: |
+     %(executable)s
+ - path: /root/startup.sh
+   permissions: '0755'
+   content: |
+     #!/bin/bash
+     cd /mnt/dirac
+     export X509_USER_PROXY=/mnt/proxy.pem
+     export PILOT_UUID="cloud://$(cat /var/lib/cloud/data/instance-id)"
+     bash /mnt/run_pilot.sh &> /mnt/dirac/startup.log
+
+yum_repos:
+  cernvm:
+    name: CernVM packages
+    baseurl: http://cvmrepo.web.cern.ch/cvmrepo/yum/cvmfs/EL/$releasever/$basearch/
+    enabled: true
+    gpgcheck: true
+    gpgkey: file:///etc/pki/rpm-gpg/RPM-GPG-KEY-CernVM
+  epel:
+    name: Extra Packages for Enterprise Linux 7 - $basearch
+    baseurl: http://download.fedoraproject.org/pub/epel/7/$basearch
+    mirrorlist: https://mirrors.fedoraproject.org/metalink?repo=epel-7&arch=$basearch
+    failovermethod: priority
+    enabled: true
+    gpgcheck: true
+    gpgkey: file:///etc/pki/rpm-gpg/RPM-GPG-KEY-EPEL-7
+
+packages:
+ - cvmfs
+ - singularity
```

Comment:
The current template contains also a hook to add extra packages and an ssh user for debugging access. 
Why this was dropped ?
```suggestion
 - singularity
 %(extra-packages)s

%(ssh-user)s
```

Reply:
We dropped all of the customisation out from the template as we only really intend it to be used as an example and for basic tests (we also dropped the CVMFS proxy setting so it should default to autodetecting a local mirror)... I'd prefer to keep this file as simple (and unchanging) as possible and then for admins to make a copy which they can customise it for their own needs (we provide a hook to point at a different file).

If we did want to include customisations, I'd rather think up a more generic way of doing this... I've got some ideas but I suspect we'd need to include pyyaml in our deps to make the patching more reliable (which may not be a bad idea for injecting the proxy & pilot script anyway: Adding the correct amount of whitespace before strings has always felt fragile).

On the subject of extra-packages: I'm very keen to encourage people to use singularityCE for the proxy isolation, in which case the package list here has no effect for the user jobs. There is also a scalability issue, we've found installing things in this way puts a lot of load on the cloud hypervisors and mirrors (it'd be better if we didn't install or update anything and just had an up-to-date image available: This has the advantage of booting the latest kernel, which is another weakness of using the default template).

Out of interest, why do you add the user this way, rather than use the default user + ssh-key deployed via the cloud infrastructure (e.g. the Instance_SSHKey option) for debugging?

Reply:
About images and extra packages. It is good to have up-to-date images with all the needed packages to be readily available at each and every cloud site. But in practice this is nearly impossible. What is more likely is to have an up-to-date basic OS image, e.g. CentOS7 official, this we can rely on,  in the end this is the site responsibility to keep them up to date. So, we then have to start from raw images and then add more necessary stuff, e.g. CVMFS and get up-to-date kernel ready for use by us. This takes longer to start, true, but it is safer and more generic across sites. In any case those pilots are meant to run for a long time, so the relative overhead should not be very important.

In our case (EGI) we have different packages to be added for different VOs. This is configurable in VMDIRAC. I think, this can be done similarly here. 

As for the user for debug access. My understanding was that the Instance_SSHKey option adds the authorized key for the root user. So, you have to login as root. Not all images allow for that, e.g. CentOS. May be I am missing something here. The suggested method certainly works, as it works in VMDIRAC.

Reply:
We've added a hook for adding packages to the list (and switched to pyyaml to make this whole process more reliable). We start from raw images too, but rely on the container to provide the packages required (so that the pilot proxy is isolated from the user jobs). Updating the kernel doesn't really work here unless you're also modifying the script to reboot the instance and continue setting up after that?

The Instance_SSHKey will generally be provisioned for whichever user is set as the default (it's a cloud-init feature): Normally this is by the image builder to be the standard user, for example on a plain upstream CentOS image, this will be added for the "centos" user.

Reply:
I do not mean that we should update the kernel while the bootstrap. I just mean that the site admins are responsible to keep the images up-to-date and we can just rely on that.